c/stcbackground.c

```diff
@@ -0,0 +1,98 @@
+#include "logging.h"
+#include "stcbackground.h"
+
+#define STC_BG_ENTRIES 101
+#define STC_BG_INTERVAL 10
+
+
+static int processInterval(STCBase *stcBase, STCModule *module, int selectStatus) {
+  STCCallbackList* moduleData = (STCCallbackList*)(module->data);
+  STCIntervalCallbackData* callbackList = moduleData->callbackList;
+
+  for (int i = 0; i < STC_BG_ENTRIES; i++) {
+    STCIntervalCallbackData* t = &callbackList[i];
+    if (t->id != -1)  /* initialise */ {
+      
+      // giving ability to disable job by making timeInterval less than zero
+      // as we return callbackdata, user can make timeInterval zero
+      if(t->timeInterval<0) continue;
+
+      t->countInterval+=STC_BG_INTERVAL;
+      if (t->countInterval <= 0) {
+        STCIntervalCallback callback = t->callback;
+        callback(moduleData->server, stcBase, module, t, t->userData);
+        t->countInterval = -t->timeInterval;
+      }
+    } else {
+      break;
+    }
+  }
+  return 0;
+};
+
+static int nextSlot(STCIntervalCallbackData* callbackList) {
+  for (int i = 0; i < STC_BG_ENTRIES; i++) {
+    if (callbackList[i].id == -1)  /* initialise */ {
+      return i;
+    }
+  }
+  return -1;
+};
+
+STCModule* stcInitBackgroundModule(void* server, STCBase *stcBase) {
+  STCCallbackList* moduleData = (STCCallbackList*)safeMalloc(sizeof(STCCallbackList),"STCCallbackList");
+  if (moduleData == NULL) {
+    return NULL;
+  }
+
+  STCIntervalCallbackData* callbackList = (STCIntervalCallbackData*)safeMalloc(sizeof(STCIntervalCallbackData)*STC_BG_ENTRIES,"STCIntervalCallbackData");
+  if (callbackList == NULL) {
+    return NULL;
+  }
+
+  for (int i = 0; i < STC_BG_ENTRIES; i++) {
+    callbackList[i].id = -1;
+  }
+  moduleData->callbackList = callbackList;
+  moduleData->server = server;
+
+  STCModule* stcBackgroundModule=stcRegisterModule(
+    stcBase,
+    STC_MODULE_BACKGROUND,
+    moduleData,
+    NULL,
+    NULL,
+    NULL,
+    processInterval
+  );
+  return stcBackgroundModule;
+};
+
+int stcAddIntervalCallback(STCModule *module, STCIntervalCallback callback, char* callbackLabel, int timeInterval, void* userData) {
+  STCCallbackList* moduleData = (STCCallbackList*)(module->data);
+  STCIntervalCallbackData* callbackList = moduleData->callbackList;
+
+  int len;
+  int slotId = nextSlot(callbackList);
+  if (slotId >= 0 && slotId<STC_BG_ENTRIES) {
+    callbackList[slotId].id = slotId;
+    len = strlen(callbackLabel);
+    if (len > STC_BG_CB_LABEL_LEN-1) {
```

Comment:
Is a label supposed to uniquely identify a callback? If yes, what happens if a label is too big, it gets truncated and you potentially have 2 identical labels? If no, what does uniquely identifies callbacks then? A label seems like a good choice for such purpose, but then we need to make sure there are no duplicate labels and if there is a duplicate we need to return an error code.

Reply:
label can be used for logging and debugging, just human readable string.
should be unique, but wont harm anything if duplicated
Will check for uniqueness while iterating for empty slot

Reply:
add label duplication check, and improved on truncation logic, thanks!